### PR TITLE
Cleanup the sources in api some

### PR DESCRIPTION
**Pull request by dwijnand**
Changes include:
* make IndexOrNth sealed
* make things final in Cmp
* align Size ascii art
* order interfaces so they read in as a sequential narrative (without using anything that's defined later)
* drop implicitly in favour of ?
* drop the unneeded self type in ScalaLib
* alphabetically sorting

### CODE REVIEW

**Comment by dwijnand** [comment]
Oh and I dropped foreach in Each (it's inherited by its parent interface Foreach)

File: api/src/main/scala/TypeClasses.scala [hunk 1]
```diff
@@ -59,13 +59,13 @@ object Pair {
...
     def apply[R, A, B](f: (A, B) => R, l: R => A, r: R => B): Cleave[R, A, B] = new Cleave[R, A, B] {
-      def split(x: R): A -> B = ((l(x), r(x)))
+      def split(x: R): A -> B = (l(x), r(x))
       def join(x: A, y: B): R = f(x, y)
     }
   }
   object Split {
     def apply[R, A, B](l: R => A, r: R => B): Split[R, A, B] =
-      new Split[R, A, B] { def split(x: R): A -> B = ((l(x), r(x))) }
+      new Split[R, A, B] { def split(x: R): A -> B = (l(x), r(x)) }
```
Comment (paulp) [inline api/src/main/scala/TypeClasses.scala]: It's no big deal, but for the record I do these double-parentheses on purpose to emphasize the tupling. Historically there have been more concrete reasons, but I think running under -Yno-adapted-args makes it mostly unnecessary.

**Comment by paulp** [comment]
LGTM. It's not easy to make a number of small changes to my code and not have me dispute the legitimacy of at least half, so nice job. Thanks.

File: api/src/main/scala/TypeClasses.scala (hunk 1 above)
Comment (dwijnand) [inline api/src/main/scala/TypeClasses.scala]: I got some recent experience with this.

Under `-Yno-adapted-args` you need to double-parentheses when, for instance, instantiating a `Future[(Int, Int)]`:

```scala
Future successful ((1, 2))
```

but when returning a tuple like here you don't.

**Pull request merged**